## Final Project Submission

Please fill out:
* Student name: Calvine Dasilver
* Student pace: full time
* Scheduled project review date/time: 
* Instructor name: Nikita
* Blog post URL:


 #  Demystifying House Sales Analysis with Regression Modeling in a Northwestern County

   ## Project Overview

   ### <li>**Business Understanding**

The real estate market plays a crucial role in the economic health and stability of a region. Understanding the factors that influence house prices is essential for both buyers and sellers to navigate the market effectively. This project focuses on a specific northwestern county in the United States, aiming to shed light on the key determinants of property valuation in this area.
   ##### Problem Statements:
What are the most significant factors influencing house prices in this northwestern county?How can we quantify the relationship between these factors and property value?Can we develop a reliable model to predict house prices based on relevant characteristics?
   ##### Challenges:
* Real estate data can be complex and multifaceted, encompassing various property features and local market trends.
* Accurately identifying and quantifying the relative impact of each factor on house prices can be challenging.
* External factors like economic conditions and interest rates might also influence prices, requiring careful consideration.

##### Proposed Solutions:
We propose utilizing multiple linear regression, a powerful machine learning technique. This method allows us to analyze a large dataset of house sales and identify the statistical relationships between various property features (e.g., square footage, number of bedrooms, location) and the corresponding sale prices.
   ##### Objectives:
1. Develop a robust multiple linear regression model that accurately predicts house prices in the chosen northwestern county.
2. Identify the most significant factors influencing property value within this specific market.
3. Provide valuable insights into the housing market dynamics of the region, benefiting potential buyers, real estate agents, and other stakeholders.
   
 
**Research questions that would help to achieve the objectives**:

1. How does the number of bedrooms, bathrooms, grade and square footage of a house correlate with its sale price in King County?
2. How much can a homeowner expect the value of their home to increase after a specific renovation project?
3. Which renovation projects have the most significant impact on a home's market value in the northwestern county?
4. Are there specific combinations of renovation projects that provide an interdependent effect on a home's market value?

   ### <li> **Data Understanding**

Our analysis leverages the King County House Sales dataset - a rich resource containing over 21,500 records and 20 distinct features(columns). Spanning house sales from May 2014 to May 2015, this dataset provides a comprehensive snapshot of the King County housing market during that period.

**The King County House Sales dataset contains the following columns;**

id - unique identified for a house

date - Date house was sold 

Price - Sale price (prediction target)

bedrooms - Number of bedrooms,

bathrooms - Number of bathrooms,

sqft_living - Square footage of living space in the home,

sqft_lot - Square footage of the lot,

floors - Number of floors (levels) in house,

view - Quality of view from house,

condition - How good the overall condition of the house is. Related to maintenance of house,

grade - Overall grade of the house. Related to the construction and design of the house,

sqft_above - Square footage of house apart from basement,

sqft_basement - Square footage of the basement,

yr_built - Year when house was built,

yr_renovated - Year when house was renovated,

zipcode - ZIP Code used by the United States Postal Service,

sqft_living15 - The square footage of interior housing living space for the nearest 15 neighbors,

sqft_lot15 - The square footage of the land lots of the nearest 15 neighbors, and

sell_yr - Date house was sold.


We need to be aware of certain constraints within the data, as these might influence our analysis and interpretation of the results. From the sources;

1. The data may contain anomalies or inconsistencies that require careful examination during analysis. For instance, a record lists a house with 33 bedrooms, which appears to be an outlier

2. It's important to consider the time frame of the data (May 2014 - May 2015) as it may not fully capture the current market dynamics in King County.
3. It's important to acknowledge the scope of the data. While it provides details on house features, it may not capture external factors such as interest rates or the overall economic climate, which can also play a role in determining property values.

DATA PREPARATION

In [1]:
import csv
import pandas as pd
import sqlite3
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from statsmodels.api import OLS
%matplotlib inline

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_validate
from sklearn.preprocessing import LabelEncoder

In [2]:
df = pd.read_csv("data/kc_house_data.csv")
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0      5650     1.0        NaN  NONE  ...      7 Average       1180   
1      7242     2.0         NO  NONE  ...      7 Average       2170   
2     10000     1.0         NO  NONE  ...  6 Low Average        770   
3      5000     1.0         NO  NONE  ...      7 Average       1050   
4      8080     1.0         NO  NONE  ...         8 Good       1680   

   sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0            0.0     1955           0.0    98178  47.5112 -122.257   
1          400.0     1951        1991.0    98125  47.7210 -122.319   
2            0.0     1933           NaN    98028  47.7379 -122.233   
3          910.0     1965           0.0    98136  47.5208 -122.393   
4            0.0     1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 21 columns]

In [3]:
print(df.columns)

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')


In [4]:
#checking null values
df.isna().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

In [16]:
#checking on data information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        21597 non-null  float64
 1   bedrooms     21597 non-null  int64  
 2   bathrooms    21597 non-null  float64
 3   sqft_living  21597 non-null  int64  
 4   condition    21597 non-null  int64  
 5   yr_built     21597 non-null  int64  
 6   sell_yr      21597 non-null  int64  
 7   house_age    21597 non-null  int64  
dtypes: float64(2), int64(6)
memory usage: 1.3 MB


In [15]:
#statistical summary
df.describe()

price      bedrooms     bathrooms   sqft_living     condition  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  21597.000000   
mean   5.402966e+05      3.373200      2.115826   2080.321850      3.409825   
std    3.673681e+05      0.926299      0.768984    918.106125      0.650546   
min    7.800000e+04      1.000000      0.500000    370.000000      1.000000   
25%    3.220000e+05      3.000000      1.750000   1430.000000      3.000000   
50%    4.500000e+05      3.000000      2.250000   1910.000000      3.000000   
75%    6.450000e+05      4.000000      2.500000   2550.000000      4.000000   
max    7.700000e+06     33.000000      8.000000  13540.000000      5.000000   

           yr_built       sell_yr     house_age  
count  21597.000000  21597.000000  21597.000000  
mean    1970.999676   2014.322962     43.323286  
std       29.375234      0.467619     29.377285  
min     1900.000000   2014.000000     -1.000000  
25%     1951.000000   2014.000000     18.000000  
50%     1975.000000   2014.000000     40.000000  
75%     1997.000000   2015.000000     63.000000  
max     2015.000000   2015.000000    115.000000

Statistical summary observation count for each column is 21597 this shows that we dont have missing values The mean value of the house price is USD 540297 while the minimum house price is USD 78000 and maximum house price is USD 7700000 The standard deviation of the house price stands at USD 367368.

In [7]:
#we'll focus on the columns mentioned above and drop the rest
df.drop(columns=['id','lat','long','sqft_lot','floors','waterfront','view','zipcode','sqft_living15','sqft_lot15','grade','yr_renovated','sqft_above','sqft_basement'],inplace=True)
df.head()

date     price  bedrooms  bathrooms  sqft_living  condition  yr_built
0  10/13/2014  221900.0         3       1.00         1180    Average      1955
1   12/9/2014  538000.0         3       2.25         2570    Average      1951
2   2/25/2015  180000.0         2       1.00          770    Average      1933
3   12/9/2014  604000.0         4       3.00         1960  Very Good      1965
4   2/18/2015  510000.0         3       2.00         1680    Average      1987

In [8]:
#checking for for null values
df.isna().sum()

date           0
price          0
bedrooms       0
bathrooms      0
sqft_living    0
condition      0
yr_built       0
dtype: int64

In [9]:
#converting date to datetime format
df['date']=pd.to_datetime(df['date'])
#extracting year from date column
df.date=df['date'].dt.year
df.head()

date     price  bedrooms  bathrooms  sqft_living  condition  yr_built
0  2014  221900.0         3       1.00         1180    Average      1955
1  2014  538000.0         3       2.25         2570    Average      1951
2  2015  180000.0         2       1.00          770    Average      1933
3  2014  604000.0         4       3.00         1960  Very Good      1965
4  2015  510000.0         3       2.00         1680    Average      1987

In [10]:
# Creating a new column for sell year
df['sell_yr'] = pd.to_datetime(df['date'],format='%Y').dt.year
df.drop(columns='date', inplace=True)
df.head()

price  bedrooms  bathrooms  sqft_living  condition  yr_built  sell_yr
0  221900.0         3       1.00         1180    Average      1955     2014
1  538000.0         3       2.25         2570    Average      1951     2014
2  180000.0         2       1.00          770    Average      1933     2015
3  604000.0         4       3.00         1960  Very Good      1965     2014
4  510000.0         3       2.00         1680    Average      1987     2015

In [11]:
#creating column house age at year of sale
df['house_age']=df['sell_yr']-df['yr_built']
df.head()

price  bedrooms  bathrooms  sqft_living  condition  yr_built  sell_yr  \
0  221900.0         3       1.00         1180    Average      1955     2014   
1  538000.0         3       2.25         2570    Average      1951     2014   
2  180000.0         2       1.00          770    Average      1933     2015   
3  604000.0         4       3.00         1960  Very Good      1965     2014   
4  510000.0         3       2.00         1680    Average      1987     2015   

   house_age  
0         59  
1         63  
2         82  
3         49  
4         28

In [12]:
#transformation of categorical values 'condition' column
df['condition'].replace(to_replace=['Poor', 'Fair', 'Average', 'Good', 'Very Good'], value=[1, 2, 3, 4, 5], inplace=True)
df.head()

price  bedrooms  bathrooms  sqft_living  condition  yr_built  sell_yr  \
0  221900.0         3       1.00         1180          3      1955     2014   
1  538000.0         3       2.25         2570          3      1951     2014   
2  180000.0         2       1.00          770          3      1933     2015   
3  604000.0         4       3.00         1960          5      1965     2014   
4  510000.0         3       2.00         1680          3      1987     2015   

   house_age  
0         59  
1         63  
2         82  
3         49  
4         28

In [13]:
#converting the 'sell_yr' column to int64
df['sell_yr'] = df['sell_yr'].astype('int64')
df.head()

price  bedrooms  bathrooms  sqft_living  condition  yr_built  sell_yr  \
0  221900.0         3       1.00         1180          3      1955     2014   
1  538000.0         3       2.25         2570          3      1951     2014   
2  180000.0         2       1.00          770          3      1933     2015   
3  604000.0         4       3.00         1960          5      1965     2014   
4  510000.0         3       2.00         1680          3      1987     2015   

   house_age  
0         59  
1         63  
2         82  
3         49  
4         28

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        21597 non-null  float64
 1   bedrooms     21597 non-null  int64  
 2   bathrooms    21597 non-null  float64
 3   sqft_living  21597 non-null  int64  
 4   condition    21597 non-null  int64  
 5   yr_built     21597 non-null  int64  
 6   sell_yr      21597 non-null  int64  
 7   house_age    21597 non-null  int64  
dtypes: float64(2), int64(6)
memory usage: 1.3 MB
